amethyst_renderer/src/memory.rs

```diff
@@ -0,0 +1,151 @@
+use std::sync::Arc;
+use std::sync::atomic::{AtomicUsize, Ordering};
+
+use gfx_hal::buffer::{CreationError, Usage, ViewError};
+use gfx_hal::device::{BindError, OutOfMemory};
+use gfx_hal::mapping::Error as MappingError;
+use gfx_hal::memory::Properties;
+use gfx_hal::{Device as GfxDevice, MemoryType};
+
+use back::*;
+
+error_chain! {
+    foreign_links {
+        BindError(BindError);
+        ViewError(ViewError);
+    }
+
+    errors {
+        InvalidAccess {
+            description("Invalid access"),
+            display("Invalid access"),
+        }
+        OutOfBounds {
+            description("Out of bounds"),
+            display("Out of bounds"),
+        }
+        OutOfMemory {
+            description("Out of memory"),
+            display("Out of memory"),
+        }
+        CreationError(inner: CreationError) {
+            description("Buffer creation error"),
+            display("Can't create buffer: {:?}", inner),
+        }
+    }
+}
+
+impl From<MappingError> for ErrorKind {
+    fn from(error: MappingError) -> ErrorKind {
+        match error {
+            MappingError::InvalidAccess => ErrorKind::InvalidAccess,
+            MappingError::OutOfBounds => ErrorKind::OutOfBounds,
+            MappingError::OutOfMemory => ErrorKind::OutOfMemory,
+        }
+    }
+}
+
+impl From<MappingError> for Error {
+    fn from(error: MappingError) -> Error {
+        ErrorKind::from(error).into()
+    }
+}
+
+impl From<OutOfMemory> for Error {
```

Comment:
Shouldn't this use `links` in the `error-chain` macro?

Reply:
`OutOfMemory` isn't `Error`
links are for other `error-chain`s. For non-chain errors there is `foreign_links`. But it still required to implement `Error`

Reply:
And `OutOfMemory` doesn't implement `Error`?

Reply:
Yeap